### PR TITLE
Scale network graph based on time interval

### PR DESCRIPTION
**Pull request by RandyMcMillan**
The Network Graph is updated to scale based on the time interval selected by the QSlider.
The graph now displays a sample history that is retained when the QSlider value is changed.
Fixes issue #532

### CODE REVIEW

**Comment by RandyMcMillan** [comment]
This is a short screen cast of the functionality.

https://user-images.githubusercontent.com/152159/151306403-456dee98-ccfa-4db4-90d6-8687936e8180.mov

**Comment by RandyMcMillan** [comment]
Here it is retaining 5 1/2 hours of samples - still snappy. 


https://user-images.githubusercontent.com/152159/151371047-ac788b14-cb5f-4bc1-b7be-ca662cdc1bdf.mov

**Comment by RandyMcMillan** [comment]
Original issue description #7481 (2016)

https://github.com/bitcoin/bitcoin/issues/7481

**Comment by RandyMcMillan** [comment]
Commit https://github.com/bitcoin-core/gui/pull/539/commits/42922129d1d8232fceff271c753e86d5ff555d27
takes the relevant samples from the global sample list to determine tmax - which is used to calculate the y axis'.
 
https://user-images.githubusercontent.com/152159/151503461-d9682cce-be68-4d12-b4f1-e563077c1338.mp4

I will remove https://github.com/bitcoin-core/gui/pull/539/commits/86120e90750347a97e4180f8db69ac793d86a641 when squashing the PR.

**Comment by RandyMcMillan** [comment]
Implemented PR #473 Log Scale Toggle


https://user-images.githubusercontent.com/152159/151606196-3981d2df-e0eb-476b-85f6-8fd6c0c07054.mp4

**Comment by rsafier** [comment]
tested it out and seemed to behave as expected.

**Comment by jonatack** [comment]
Concept ACK, thanks for tackling this. Will review/test shortly.

**Comment by RandyMcMillan** [comment]
commit 83e521c - attaches the scale toggle to a mouse double click event opposed to a single click event.
Toggling the scale is more intentional this way - plus it make a single click event available for something more useful.

**Comment by RandyMcMillan** [comment]
In some of the previous screen shots - the colors look muddy - commit 8eeb1f1 adjusts the visual output to avoid the way the colors were mixing and creating a muddy effect. 


https://user-images.githubusercontent.com/152159/151662598-84a5b3ef-8f0b-4c40-91c4-95e3b3012e4e.mp4

**Comment by RandyMcMillan** [comment]
This PR feels complete - and stands on its own...
Thanks for the review/feedback!

Screen Shot 2022-01-29 at 2 10 01 PM (2)

**Comment by w0xlt** [comment]
Concept ACK

File: src/qt/trafficgraphwidget.cpp [hunk 1]
```diff
@@ -47,30 +49,43 @@ int TrafficGraphWidget::getGraphRangeMins() const
...
+
 void TrafficGraphWidget::paintEvent(QPaintEvent *)
 {
     QPainter painter(this);
     painter.fillRect(rect(), Qt::black);
 
     if(fMax <= 0.0f) return;
 
-    QColor axisCol(Qt::gray);
+    QColor axisCol(QColor(78,78,78,255));
```
Comment (rebroad) [inline src/qt/trafficgraphwidget.cpp:88]: ?

File: src/qt/trafficgraphwidget.cpp [hunk 2]
```diff
@@ -135,32 +161,35 @@ void TrafficGraphWidget::updateRates()
     nLastBytesIn = bytesIn;
     nLastBytesOut = bytesOut;
 
-    while(vSamplesIn.size() > DESIRED_SAMPLES) {
+    while(vSamplesIn.size()  > FLT_MAX - 1) {
```
Comment (rebroad) [inline src/qt/trafficgraphwidget.cpp:164]: how much memory does this need?

Comment (RandyMcMillan) [reply src/qt/trafficgraphwidget.cpp:164]: I agree - I have been thinking about how to do this. Maybe pruning the list at larger intervals - keeping samples that are above a moving average. Open to suggestions on implementation. :) 

Thanks for looking at this!

Comment (RandyMcMillan) [reply src/qt/trafficgraphwidget.cpp:164]: We can chunk the list at DESIRED_SAMPLES intervals - calculate the moving average - pruning samples that fall below it.

File: src/qt/trafficgraphwidget.cpp (hunk 1 above)
Comment (RandyMcMillan) [reply src/qt/trafficgraphwidget.cpp:88]: I didn't check to see if there was a Qt defined dark gray - It is ideal to let the samples stand out above the scale lines - we can tweak this. Note: there is plenty of room unused in the side panel - I have some code laying around to define user preferred color palettes - could be a useful feature for usability/accessibility.

Comment (RandyMcMillan) [reply src/qt/trafficgraphwidget.cpp:88]: A thought was to implement Solarized Light/Dark color palettes as user options - I am a fan - and plenty of research has gone into these color schemes.

**Comment by RandyMcMillan** [comment]
To be clear - This PR includes a proposal of a redesign (with considerations for usability/accessibility/etc) - which I am not emotionally attached to - we can always select the commits that directly fix the original bug - and merge a separate PR with limited scope - I just think it is useful to implement a fix with a broader context in mind. 😄

**Comment by RandyMcMillan** [comment]
rebase to fix Centos CI issue
[874b2d8](https://github.com/bitcoin-core/gui/pull/539/commits/874b2d8ced4e63fb4a2d95951373e1a845d7231b) - add hide panel logic which collapses window to a "desktop widget"
[bb5b4cc](https://github.com/bitcoin-core/gui/pull/539/commits/bb5b4cc41c700dc1a9c97f8e112e183f436f3ef1) - changed DESIRED_SAMPLES to 400 - reducing overall memory usage - going to try some other ideas to minimize memory usage. (interim adjustment based on @rebroad's comment)
[c0f46d2](https://github.com/bitcoin-core/gui/pull/539/commits/c0f46d293a283fd0bbcab168ff7e2ebfc383d81b) - the linear scale is the default presentation - this is familiar to users - a follow up PR should include a note on adding double click log scale to release notes.


commit 874b2d8 screen cast: 


https://user-images.githubusercontent.com/152159/152240134-72f51e3b-697c-4065-a752-6108c71b8ac9.mp4

**Comment by rebroad** [comment]
I've been working on this, and I have a pull request almost ready to make - storing 600 samples per time range, and 14 ranges in total (from 1 minute to 28 days), so 8400 samples in total.

https://user-images.githubusercontent.com/1530283/152860520-9aee9d52-a13b-44bd-bac9-2ec9f143a769.mp4

**Comment by RandyMcMillan** [comment]
@rebroad - awesome!

**Review by CryptoGodfatherVA4 — APPROVED** [review]

**Comment by RandyMcMillan** [comment]
@rebroad - I am curious if your memory managment is ready...

I am happy to merge it into this PR...

**Comment by rebroad** [comment]
> What is the status of this PR?

I'm currently rebasing it. Should update soon.

**Comment by rebroad** [comment]
> There hasn't been much activity lately and the patch still needs rebase. What is the status here?
> 
> * Is it still relevant? ➡️ Please solve the conflicts to make it ready for review and to ensure the CI passes.
> * Is it no longer relevant? ➡️ Please close.
> * Did the author lose interest or time to work on this? ➡️ Please close it and mark it 'Up for grabs' with the label, so that it can be picked up in the future.

How do I label it as up for grabs?

**Comment by hebasto** [comment]
> How do I label it as up for grabs?

Done.

**Pull request closed without merging**